prow/cmd/webhook-server/main.go

```diff
@@ -48,47 +47,125 @@ type ClientInterface interface {
+	optionGroup := []flagutil.OptionGroup{&o.kubernetes}
+	optionGroup[0].AddFlags(fs)
+	fs.Parse(args)
+	return o
+}
+
 func main() {
 	logrusutil.ComponentInit()
 	logrus.SetLevel(logrus.DebugLevel)
-	p := flag.String("project-id", 
-	"", "Project ID for storing GCP Secrets")
-	e := flag.Int("expiry-years", 30, "CA certificate expiry in years")
-	dns := flagutil.NewStrings("validation-webhook-service", "validation-webhook-service.default", "validation-webhook-service.default.svc")
+	dns := prowflagutil.NewStrings("validation-webhook-service", "validation-webhook-service.default", "validation-webhook-service.default.svc")
 	flag.Var(&dns, "dns", "DNS Names CA-Cert config")
 	flag.Parse()
-	exp := *e
-	projectId := *p
-	if len(projectId) == 0 {
-		logrus.Fatal("project-id flag not supplied")
+	o := gatherOptions(flag.NewFlagSet(os.Args[0], flag.ExitOnError), os.Args[1:]...)
+	o.dnsNames = dns.Strings()
+	if err := o.Validate(); err != nil {
+		logrus.WithError(err).Fatal("Invalid options")
 	}
-	client, err := secretmanager.NewClient(projectId, false)
+	kubeCfg, err := o.kubernetes.InfrastructureClusterConfig(o.dryRun)
 	if err != nil {
-		logrus.WithError(err).Fatal("Unable to create secret manager client")
+		logrus.WithError(err).Fatal("Error getting kubeconfig")
 	}
+	var certFile string
+	var privKeyFile string
+	var tempDir string
 	ctx := context.Background()
-	cert, privKey, err := getGCPSecrets(client, ctx, exp, dns)
+	cl, err := ctrlruntimeclient.New(kubeCfg, ctrlruntimeclient.Options{})
+	if err != nil {
+		logrus.WithError(err).Fatal("Could not create writer client")
+	}
+	var client ClientInterface
+	if len(o.projectId) > 0 {
+		client, err = secretmanager.NewClient(o.projectId, false)
+		if err != nil {
+			logrus.WithError(err).Fatal("Unable to create secret manager client")
+		}
+		tempDir, certFile, privKeyFile = handleSecrets(client, ctx, o, cl)
+		defer os.RemoveAll(tempDir)
+	}
+	if len(o.fileSystemPath) > 0 {
+		absPath, err := filepath.Abs(o.fileSystemPath)
+		if err != nil {
+			logrus.WithError(err).Fatal("Unable to generate absolute file path")
+		}
+		client = &FileSystemClient{
+			path:   absPath,
+			expiry: o.expiryInYears,
+			dns:    o.dnsNames,
+		}
+		tempDir, certFile, privKeyFile = handleSecrets(client, ctx, o, cl)
+		defer os.RemoveAll(tempDir)
+	}
+	http.HandleFunc("/validate", serveValidate)
+	logrus.Info("Listening on port 8008...")
+	logrus.Fatal(http.ListenAndServeTLS(":8008", certFile, privKeyFile, nil))
+}
+
+func handleSecrets(client ClientInterface, ctx context.Context, o options, cl ctrlruntimeclient.Client) (string, string, string) {
+	cert, privKey, _, err := getGCPSecrets(client, ctx, o.expiryInYears, o.dnsNames)
+	var caPem string
 	if err != nil && strings.Contains(err.Error(), gcpSecretError) {
-		logrus.Infof("%v, Will now proceed to create GCP Secret", err)
-		cert, privKey, err = createGCPSecret(client, ctx, exp, dns.Strings())
+		logrus.Infof("%v, Will now proceed to create secrets", err)
```

Comment:
structured logs is generally preferred
```suggestion
		logrus.WithError(err).Info("Secret exists, skip creating secret")
```